File: 230.Kth-Smallest-Element-in-a-BST/tree.c
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int count(struct TreeNode *root) {
    if (!root) {
        return 0;
    }

    // O(n)
    int l = count(root->left);
    int r = count(root->right);

    return 1 + l + r;
}

int kthSmallest(struct TreeNode *root, int k) {
    if (!root) {
        return -1;
    }
    
    // O(n)
    int left = 1 + count(root->left);

    if (left < k) {
        // O(n^2)
        return kthSmallest(root->right, k - left);
    }

    if (k < left) {
        // O(n^2)
        return kthSmallest(root->left, k);
    }

    return root->val;
}

File: 1448.Count-Good-Nodes-in-Binary-Tree/tree.rs
use std::{cell::RefCell, rc::Rc};

impl Solution {
    fn __good_nodes(root: Option<Rc<RefCell<TreeNode>>>, cap: i32) -> i32 {
        root.map(|mut node| {
            let mut borrow = node.borrow_mut();
            let new_cap = cap.max(borrow.val);

            // O(n)
            let l = Self::__good_nodes(borrow.left.take(), new_cap);
            let r = Self::__good_nodes(borrow.right.take(), new_cap);
            let this = (borrow.val >= cap) as i32;

            l + this + r
        })
        .unwrap_or(0)
    }

    pub fn good_nodes(root: Option<Node>) -> i32 {
        // O(n)
        match root.as_ref().map(|node| node.borrow().val) {
            Some(val) => Self::__good_nodes(root, val),
            None => 0,
        }
    }
}
